### PR TITLE
fix(spec): Harden decimal and fixed type string parsing

### LINKED ISSUE #3229: Schema data type parsing needs hardening
### Apache Iceberg Rust version

main

### Describe the bug

Today, the data type parsing can parse some really weird combinations such as `decimal(5, 8))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))`.

This is not a big issue, but I think it would make more sense for us to parse a stricter subset - not overly strict, but one that is predictable.

### To Reproduce

_No response_

### Expected behavior

We should reject types that contain additional `)` or otherwise appears malformed.

I'd recommend we adopt a regex pattern, similar to iceberg-go. https://github.com/apache/iceberg-go/blob/8832cf696b5cc2a54767b5f2fa4c06b138c04495/types.go#L35-L40

### Willingness to contribute

I can contribute a fix for this bug independently

### PR DESCRIPTION
**Pull request by shoemoney**
## Which issue does this PR close?

- Closes #3229.

## What changes are included in this PR?

`deserialize_decimal` (crates/iceberg/src/spec/datatypes.rs:350-363 on main) and
`deserialize_fixed` (:376-387 on main) used `trim_start_matches` / `trim_end_matches`.
Those strip *repeated* occurrences and never require the prefix to be present, so the
following all parsed successfully before this change:

- `"decimal(5, 2)))))"` -> `Decimal { precision: 5, scale: 2 }`
- `"decimal(decimal(5, 2)))"` -> `Decimal { precision: 5, scale: 2 }`
- `"decimal(5, 2"` (no closing paren) -> `Decimal { precision: 5, scale: 2 }`
- `"fixed[fixed[16]]]"` -> `Fixed(16)`
- `"fixed[16"` -> `Fixed(16)`

Precision was also unvalidated, so `"decimal(50, 2)"` deserialized into a
`PrimitiveType::Decimal` and round-tripped back out, even though `Type::decimal` and
`Type::decimal_required_bytes` enforce `precision <= MAX_DECIMAL_PRECISION` (38). This is
reachable from every JSON schema parse (`impl Deserialize for PrimitiveType`), so a
`Schema` or `TableMetadata` written through iceberg-rust could carry a decimal type that
iceberg-java and pyiceberg reject; the error only surfaced much later and far away, in
`arrow::schema` (`validate_decimal_precision_and_scale`) or `avro::schema`. `scale > precision`
was likewise accepted.

The fix parses strictly with `strip_prefix` / `strip_suffix`, and applies the invariants the
crate already states elsewhere: `0 < precision <= MAX_DECIMAL_PRECISION` and `scale <= precision`.

Not fixed deliberately: this PR does not introduce a regex dependency as the issue suggests,
since `strip_prefix` + `split_once` covers the same grammar with no new dependency, and it does
not touch the lenient `starts_with("decimal")` dispatch in `impl Deserialize for PrimitiveType`,
which now simply routes malformed input to a clear error instead of silently accepting it.

## Are these changes tested?

Yes, two new unit tests in the existing `spec::datatypes` test module.

Before the fix (test only, source pristine):

```
test spec::datatypes::tests::test_reject_malformed_decimal_and_fixed_type_strings ... FAILED

---- spec::datatypes::tests::test_reject_malformed_decimal_and_fixed_type_strings stdout ----
panicked at crates/iceberg/src/spec/datatypes.rs:1346:13:
expected "decimal(50, 2)" to be rejected

test result: FAILED. 14 passed; 1 failed; 0 ignored; 0 measured; 1743 filtered out
```

After the fix:

```
$ cargo test -p iceberg --lib spec::
test result: ok. 475 passed; 0 failed; 0 ignored; 0 measured; 1283 filtered out

$ cargo test -p iceberg --lib
test result: ok. 1758 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

`rustfmt` (nightly-2026-04-16, per rust-toolchain.toml) and `cargo clippy -p iceberg --lib
--all-features` are both clean on the touched file.

## AI Disclosure

Written in conjunction with my pair programmer Claude.

This review follow-up was prepared with AI assistance.

### CODE REVIEW

File: crates/iceberg/src/spec/datatypes.rs [hunk 1]
```diff
@@ -350,16 +350,30 @@ impl Serialize for PrimitiveType {
...
+        .ok_or_else(|| D::Error::custom(format!("Decimal requires precision and scale: {s}")))?;
+
+    let precision: u32 = precision.trim().parse().map_err(|_| malformed())?;
+    let scale: u32 = scale.trim().parse().map_err(|_| malformed())?;
+
+    if precision == 0 || precision > MAX_DECIMAL_PRECISION {
+        return Err(D::Error::custom(format!(
+            "Decimals with precision larger than {MAX_DECIMAL_PRECISION} are not supported: {precision}"
+        )));
+    }
```
[the remark this answers is not in the document]
Comment (shoemoney) [reply crates/iceberg/src/spec/datatypes.rs]: Fixed in 7d88ecb. Zero precision now has its own error message; the regression test failed with the old message and passes with the fix.

File: crates/iceberg/src/spec/datatypes.rs [hunk 2]
```diff
@@ -1326,6 +1338,65 @@ mod tests {
...
+            (r#""fixed[16]""#, Type::Primitive(PrimitiveType::Fixed(16))),
+        ] {
+            check_type_serde_roundtrip_value(json, expected);
+        }
+    }
+
+    fn check_type_serde_roundtrip_value(json: &str, expected_type: Type) {
+        let parsed: Type = serde_json::from_str(json).unwrap();
+        assert_eq!(parsed, expected_type);
+    }
```
[the remark this answers is not in the document]
Comment (shoemoney) [reply crates/iceberg/src/spec/datatypes.rs:1440]: Fixed in 7d88ecb. The helper now serializes the expected type and deserializes it again before comparing. All 1,759 iceberg library tests, nightly formatting, and crate Clippy passed.

File: crates/iceberg/src/spec/datatypes.rs [hunk 3]
```diff
@@ -1326,6 +1343,79 @@ mod tests {
...
+            r#""decimal(decimal(5, 2)))""#,
+            r#""decimal(5, 2""#,
+            r#""decimal(5, 2)))))""#,
+            r#""decimal(5, 2, 3)""#,
+            r#""decimal(-5, 2)""#,
+            r#""decimal()""#,
+            r#""fixed[fixed[16]]]""#,
+            r#""fixed[16""#,
+            r#""fixed[16]]]""#,
+            r#""fixed[]""#,
```
Comment (mkroll-db) [inline crates/iceberg/src/spec/datatypes.rs:1370]: Suggestion for some more tests, in particular testing negative numbers (go has them excluded in the regex):

```suggestion
            r#""decimal(50, 2)""#,
            r#""decimal(0, 0)""#,
            r#""decimal(5, 8)""#,
            r#""decimal(decimal(5, 2)))""#,
            r#""decimal(decimal(5, 2)""#,
            r#""decimal(5, 2""#,
            r#""decimal(5, 2)))))""#,
            r#""decimal(5, 2, 3)""#,
            r#""decimal(-5, 2)""#,
            r#""decimal(5, -2)""#,
            r#""decimal(-5, -2)""#,
            r#""decimal(-2, -5)""#,
            r#""decimal((5, 2))""#,
            r#""decimal[5, 2]""#,
            r#""decimal()""#,
            r#""fixed[fixed[16]]]""#,
            r#""fixed[16""#,
            r#""fixed[16]]]""#,
            r#""fixed[[16]]""#,
            r#""fixed[[16]""#,
            r#""fixed(16)""#,
            r#""fixed[]""#,
```

File: crates/iceberg/src/spec/datatypes.rs [hunk 4]
```diff
@@ -1326,6 +1343,79 @@ mod tests {
...
+                    scale: 10,
+                }),
+            ),
+            (
+                r#""decimal(5,2)""#,
+                Type::Primitive(PrimitiveType::Decimal {
+                    precision: 5,
+                    scale: 2,
+                }),
+            ),
```
Comment (mkroll-db) [inline crates/iceberg/src/spec/datatypes.rs:1413]: Some more 'corner' case tests:
```suggestion
            ),
            (
                r#""decimal(5, 0)""#,
                Type::Primitive(PrimitiveType::Decimal {
                    precision: 5,
                    scale: 0,
                }),
            ),
            (
                r#""decimal(5, 5)""#,
                Type::Primitive(PrimitiveType::Decimal {
                    precision: 5,
                    scale: 5,
                }),
            ),

```

**Review by mkroll-db — APPROVED** [review]
It matches mostly the [go implementation](https://github.com/apache/iceberg-go/blob/8832cf696b5cc2a54767b5f2fa4c06b138c04495/types.go#L35-L40) and the difference in how negative numbers are handled is covered by tests.
Added some smaller suggestions to add more tests.

File: crates/iceberg/src/spec/datatypes.rs (hunk 3 above)
Comment (shoemoney) [reply crates/iceberg/src/spec/datatypes.rs:1370]: Added the missing negative, nested and delimiter cases in efb7ebbd. All 1,759 iceberg library tests pass, along with format and strict Clippy checks.

File: crates/iceberg/src/spec/datatypes.rs (hunk 4 above)
Comment (shoemoney) [reply crates/iceberg/src/spec/datatypes.rs:1413]: Added `decimal(5, 0)` and `decimal(5, 5)` round trips in efb7ebbd; both are covered by the passing library suite.

**Review by dannycjones — APPROVED** [review]
LGTM, thanks for picking this up!

**Comment by mkroll-db** [comment]
Thanks a lot.
